RFC-0039-cuda-support.md

```diff
@@ -0,0 +1,196 @@
+
+## **Proposed Policy and Process**
+
+### We should introduce new version of CUDA when
+
+- It enables new important GPU architecture (For example A100 with CUDA-11.1)
+- It brings significant performance improvements
+- It fixes significant correctness issues
+- It significantly reduces binary/memory footprint
+
+### We would deprecate version of CUDA when
+
+As soon as we introduce a new Experimental Version we should consider moving the previous Experimental Version to Stable, and decommission the previous Stable version. Typically we want to support 2-3 versions of CUDA as follows:
+
+- Optional Legacy Version: If we need to have 1 version for backend compatibility or to work around the current limitation. For example: CUDA older driver is incompatible with newer CUDA version
+- Stable Version: This is a stable CUDA version that is used most of the time. This is the version we want to upload to PyPI. Please note that if the stable version is equal or slightly older to the version then the version fbcode is using, we probably need to keep it in CI/CD system as Optional Legacy Version.
+- Latest Experimental Version: This is the latest version of CUDA that we want to support. Minimal requirement is to have it available in nightly releases (CD)
+
+### Detailed Process of Introducing new CUDA version
+
+1. Evaluate CUDA update necessity
+  Goal: When Any of this is true:
+  - It enables new important GPU architecture (For example A100 with CUDA-11.1)
+  - It brings significant performance improvements
+  - It fixes significant correctness issues
+  - It significantly reduces binary/memory footprint
+
+2. Evaluate if we have all packages for update
+  When: As soon as Update determined to be necessary
+  Goal: Make sure everything is available to perform complete upgrade of CUDA and dependencies
+
+3. Update CUDA in CD (this is necessary condition to qualify for CUDA version be released as Experimental)
+  When: Evaluate if we have all packages for update is complete
+  Goal: Make sure all binaries Linux/Windes on wheels and conda platforms are produced on nightly
+
+4. Update CUDA in CI
+  When: Update CUDA in CD is complete
+  Goal: Make sure CI jobs tests on new CUDA versions, all failing tests should be identified and reported
+
+5. Run benchmarks to compare new experimental CUDA version to current stable CUDA version
+  When: Update CUDA in CD  is complete
+  Goal: Make sure we identify all performance regressions
+
+
+### Detailed Process of Deprecating CUDA version
+
+1. Evaluate deprecation of legacy CUDA version from CI/CD
+When: We completed CUDA update and previous experimental CUDA version is qualified to be stable and we have 3 supported versions (legacy, stable and experimental)
+Goal: Make sure we support 2 versions of CUDA, supporting 3 versions can be an exception for certain release where we need to keep legacy version
```

Comment:
> supporting 3 versions can be an exception for certain release where we need to keep legacy version

Can you share examples of when we have needed to keep a legacy version around in the past?  Sounds like we might incur a significant CI cost to test 50% more CUDA versions

Reply:
Yes this is current situation with CUDA 11.8: https://github.com/pytorch/pytorch/issues/147383

Reply:
Let's add this as an example here

Reply:
Done